File: algorithm/dijkstra.js
//this file is used to solve dijkstra algorithm
//to find nearest distance for each vertex i,j
let vertex = 6;
let matrix = new Array(vertex);
for(let i = 0; i < vertex; i++){
	matrix[i] = new Array(vertex);
	matrix[i].fill(500);//we note 500 as maximum distance
	matrix[i][i] = 0;
}

console.log("initial matrix: ",matrix);

//add edge between u and v
//where 0 <= u,v <= vertex-1 && u != v
function addEdge(u,v,weight){
	matrix[u][v] = weight;
}

//here we have 6 vertexes, 0,1,2,3,4,5 and 9 edges
addEdge(0,1,1);
addEdge(0,2,12);
addEdge(1,2,9);
addEdge(1,3,3);
addEdge(3,2,4);
addEdge(2,4,5);
addEdge(3,4,13);
addEdge(3,5,15);
addEdge(4,5,4);

console.log("after adding edges, matrix is: ",matrix);

function dijkstra(matrix,v){
		let n = matrix.length;
		let tmpDistance = new Array(n);
		let minimized = new Array(n);
		for(let i = 0; i < n; i++){
			minimized[i] = false;
			if(i == v) minimized[i] = true;
		}//note whether arrived at minimized value
		for(let i = 0; i < n; i++){
			tmpDistance[i] = matrix[v][i];
		}//initialize distance for vertex v

		for(let i = 0; i < n - 1; i++){
			//to find nearest vertex for v
			let min = 500;
			let u;
			for(let j = 0; j < n; j++){
				if(minimized[j] == false && tmpDistance[j] < min){
					min = tmpDistance[j];
					u = j;
				}
			}
			minimized[u] = true;
			//start from beginning to update distance
			for(let k = 0; k < n; k++){
				if(matrix[u][k] < 500 && matrix[u][k] > 0){
					if(matrix[u][k] + tmpDistance[u] < tmpDistance[k]){
						tmpDistance[k] = matrix[u][k] + tmpDistance[u];
					}
				}
			}
		}
		console.log("nearest distance starting from vertex: ",v);
		console.log(tmpDistance);
}

dijkstra(matrix,0);

File: algorithm/dijkstra.cpp
//this file is used to create a graph and realize breadth first search
//as well as depth first search
//in adjacent list
#include <iostream>
#include <stdio.h>
#include <string>
#include<vector>
#include<stack>
#include<queue>
using namespace std;
class SparseGraph{
public:
	int n, m;//n nodes, m edges
	bool directed;//directed or non-directed
	vector<vector<int> > g;//2D vector

	//construct function
	SparseGraph(int n, bool directed){
		this->n = n;
		this->directed = directed;
		for(int i = 0; i < n; i++){
			g.push_back(vector<int>());
		}
		//establish matrix for storing distance
		for(int i = 0; i < n; i++){
			for(int j = 0; j < n; j++){
				if(i == j) g[i].push_back(0);
				else{
					g[i].push_back(500);//we use 500 to note maximum distance, or INT_MAX
				}
			}
		}
	}

	//deconstruct function
	~SparseGraph(){}

	//get node number
	int getV(){
		return n;
	}

	//get edge number
	int getE(){
		return m;
	}

	//add edge between two nodes
	void addEdge(int v, int w, int weight){
		g[v][w] = weight;//v -> w
		if(!directed){
			g[w][v] = weight;
		}//if non-directed, w -> v
		m++;//add an edge from v to w
	}

	void print(){
		for(int i = 0; i < g.size(); i++){
			cout<<"edge "<<i<<" : ";
			for(int j = 0; j < g[i].size(); j++){
				cout<<g[i][j]<<" ";
			}
			cout<<endl;
		}
	}

	void dijkstra(int v){
		int tmpDistance[n];
		bool minimized[n];
		for(int i = 0; i < n; i++){
			minimized[i] = false;
			if(i == v) minimized[i] = true;
		}//note whether arrived at minimized value
		for(int i = 0; i < n; i++){
			tmpDistance[i] = g[v][i];
		}//initialize distance for vertex v

		for(int i = 0; i < n - 1; i++){
			//to find nearest vertex for v
			int min = 500;
			int u;
			for(int j = 0; j < n; j++){
				if(minimized[j] == false && tmpDistance[j] < min){
					min = tmpDistance[j];
					u = j;
				}
			}
			minimized[u] = true;
			//start from beginning to update distance
			for(int k = 0; k < n; k++){
				if(g[u][k] < 500 && g[u][k] > 0){
					if(g[u][k] + tmpDistance[u] < tmpDistance[k]){
						tmpDistance[k] = g[u][k] + tmpDistance[u];
					}
				}
			}
		}
		cout<<"nearest distance starting from vertex "<<v<<" is: "<<endl;
		for(int i = 0; i < n; i++){
			cout<<tmpDistance[i]<<" ";
		}
		cout<<endl;
	}
};

int main(){
	
	int vertex,edge;
	cout<<"input vertex number"<<endl;
	cin>>vertex;
	cout<<"input edge number"<<endl;
	cin>>edge;
	SparseGraph myGraph(vertex,true);
	for(int i = 0; i < edge; i++){
		int from,to,weight;
		cout<<"add an edge and its weight: "<<endl;
		cin>>from>>to>>weight;
		myGraph.addEdge(from,to,weight);
	}

	myGraph.print();
	myGraph.dijkstra(0);
    return 0;
}
